libsodium-sys/build.rs

```diff
@@ -1,38 +1,91 @@
 extern crate bindgen;
+#[cfg(not(windows))]
+extern crate cc;
+#[cfg(not(target_env = "msvc"))]
+extern crate flate2;
+extern crate http_req;
+#[cfg(target_env = "msvc")]
+extern crate libc;
 extern crate pkg_config;
+extern crate sha2;
+#[cfg(not(target_env = "msvc"))]
+extern crate tar;
+#[macro_use]
+extern crate unwrap;
+#[cfg(target_env = "msvc")]
+extern crate vcpkg;
+#[cfg(target_env = "msvc")]
+extern crate zip;
 
+use http_req::request;
+use sha2::{Digest, Sha256};
 use std::env;
-use std::path::PathBuf;
+use std::fs;
+use std::io::Cursor;
+use std::path::{Path, PathBuf};
 
-#[cfg(target_env = "msvc")]
-extern crate vcpkg;
+static DOWNLOAD_BASE_URL: &'static str = "https://download.libsodium.org/libsodium/releases/";
+static VERSION: &'static str = "1.0.16";
+
+#[cfg(target_env = "msvc")] // libsodium-<VERSION>-msvc.zip
+static SHA256: &'static str = "0580d54f57594a7cb493607cec6e7045369fb67d43623491523781e901589948";
+
+#[cfg(all(windows, not(target_env = "msvc")))] // libsodium-<VERSION>-mingw.tar.gz
+static SHA256: &'static str = "5b81a4fc5d0de36dbda7efeaf355c133d4f6cc0b4dbf69bbe46ef7f5a6baa639";
+
+#[cfg(not(windows))] // libsodium-<VERSION>.tar.gz
+static SHA256: &'static str = "eeadc7e1e1bcef09680fb4837d448fbdf57224978f865ac1c16745868fbd0533";
 
 fn main() {
     println!("cargo:rerun-if-env-changed=SODIUM_LIB_DIR");
     println!("cargo:rerun-if-env-changed=SODIUM_INC_DIR");
     println!("cargo:rerun-if-env-changed=SODIUM_STATIC");
+    println!("cargo:rerun-if-env-changed=SODIUM_FROM_SRC");
+
     if cfg!(target_env = "msvc") {
         // vcpkg requires to set env VCPKGRS_DYNAMIC
         println!("cargo:rerun-if-env-changed=VCPKGRS_DYNAMIC");
     }
 
+    let from_src = env::var("SODIUM_FROM_SRC").is_ok();
     let include_dir = {
-        let lib_dir_isset = env::var_os("SODIUM_LIB_DIR").is_some();
-        let inc_dir_isset = env::var_os("SODIUM_INC_DIR").is_some();
-        if lib_dir_isset || inc_dir_isset {
-            find_libsodium_env()
+        if from_src {
+            Some(build_libsodium())
         } else {
-            find_libsodium_pkg()
+            let lib_dir_isset = env::var_os("SODIUM_LIB_DIR").is_some();
+            let inc_dir_isset = env::var_os("SODIUM_INC_DIR").is_some();
+            if lib_dir_isset || inc_dir_isset {
+                if from_src {
```

Comment:
This is gonna always be false since we are in `if from src {} else { HERE }`

Reply:
woop, good catch, will fix